internal/keepers/ocr.go

```diff
@@ -43,40 +40,30 @@ func (k *keepers) Observation(ctx context.Context, _ types.ReportTimestamp, _ ty
 func (k *keepers) Report(ctx context.Context, _ types.ReportTimestamp, _ types.Query, attributed []types.AttributedObservation) (bool, types.Report, error) {
 	var err error
 
-	rdm, keys, err := sortedDedupedKeyList(attributed)
+	// pass the filter to the dedupe function
+	// ensure no locked keys come through
+	keys, err := sortedDedupedKeyList(attributed, k.filter.Filter())
 	if err != nil {
 		return false, nil, fmt.Errorf("%w: failed to sort/dedupe attributed observations", err)
 	}
 
-	var sent int64
-	values := make([]int64, len(rdm))
-	for i, r := range rdm {
-		values[i] = r.Value
-		if r.Observer == k.id {
-			sent = r.Value
-		}
-	}
-
-	// like drawing straws, if the lowest value from all nodes was sent by this
-	// node, this node should be the next transmitter
-	isTransmitter := sent == lowest(values)
-	k.logger.Printf("node selected as transmitter: %t", isTransmitter)
-	k.mu.Lock()
-	k.transmit = isTransmitter
-	k.mu.Unlock()
-
 	// select, verify, and build report
 	toPerform := make([]ktypes.UpkeepResult, 0, 1)
 	for _, key := range keys {
+
 		upkeep, err := k.service.CheckUpkeep(ctx, key)
 		if err != nil {
 			return false, nil, fmt.Errorf("%w: failed to check upkeep from attributed observation", err)
 		}
 
-		if upkeep.State == ktypes.Perform {
+		if upkeep.State == ktypes.Eligible {
 			// only build a report from a single upkeep for now
 			k.logger.Printf("reporting %s to be performed", upkeep.Key)
 			toPerform = append(toPerform, upkeep)
+			err = k.filter.Add(upkeep.Key)
```

Comment:
I don't think we should add here. Should be done once a report with signatures is generated

Reply:
The reason is because query/observer/report is a different process than accept/transmit. The former has the ability to continue moving on at its own pace meaning round 2 of the former can begin before round 1 of the latter has completed. We need to be able to move through ids quickly in the first stage because we're doing them one report at a time.